File: easy/anagram-check/python/anagram_check.py
# easy/anagram-check/python/anagram_check.py
from collections import Counter


def is_anagram(s: str, t: str) -> bool:
    """
    Checks if two strings are anagrams of each other.

    Time Complexity: O(n), where n is the length of the strings.
    Space Complexity: O(k), where k is the number of unique characters.
    """
    # Anagrams must have the same length.
    if len(s) != len(t):
        return False

    # Count character frequencies and compare the counts.
    return Counter(s) == Counter(t)


# Verification block for simple testing
if __name__ == "__main__":
    print(f"'listen' and 'silent' are anagrams: {is_anagram('listen', 'silent')}")
    print(f"'rat' and 'car' are anagrams: {is_anagram('rat', 'car')}")

File: easy/anagram-check/cpp/anagram_check.cpp
#include <iostream>
#include <string>
#include <unordered_map>
using namespace std;

bool isAnagram(string s, string t) {
    if (s.length() != t.length()) {
        return false;
    }
    
    unordered_map<char, int> charCount;
    
    // Count characters in s
    for (char c : s) {
        charCount[c]++;
    }
    
    // Decrement count for characters in t
    for (char c : t) {
        charCount[c]--;
        if (charCount[c] < 0) {
            return false;
        }
    }
    
    return true;
}

int main() {
    // Test cases
    cout << boolalpha;
    cout << isAnagram("anagram", "nagaram") << endl; // true
    cout << isAnagram("rat", "car") << endl; // false
    cout << isAnagram("listen", "silent") << endl; // true
    cout << isAnagram("hello", "world") << endl; // false
    cout << isAnagram("aacc", "ccac") << endl; // false
    cout << isAnagram("railsafety", "fairytales") << endl; // true
    cout << isAnagram("dormitory", "dirtyroom") << endl; // true
    
    return 0;
}
